### PR TITLE
Add option to disable automatic resize of main window

### LINKED ISSUE #1560: Window is spontaneously moved/recreated during Yoshi's island's opening
The main emulator window suddenly teleports to the middle of the screen when Yoshi's island switches from the island scene back to the storybook scenes.

### PR DESCRIPTION
**Pull request by kalimag**
- Add option to disable the automatic resizing of the main window whenever the framebuffer size changes
  - Window can be resized arbitrarily and stays that size
  - Selecting a standard scale through the menu/keybinds/API still works
- Restore last window size after restart (if automatic resize is disabled)
- Minor changes to saved window position
  - Make config values nullable instead of magic `-1` value
  - Don't save window position if minimized (this didn't quite work as intended) or maximized

This feature is useful for streamers who want a consistent capture size, as well as for people who want to fit the BizHawk window into a fixed layout with other windows (corner snapping etc.).

This is not the same functionality as the custom resolution in the display settings, which stretches the framebuffer to the specified size and fits the window to that size exactly.

With the feature in this PR, the window/client area stays a consistent size while automatic letterboxing/pillarboxing, integer resizing, 1:1 pixels etc. still work. This is preferable for working with multiple games, resolution changes mid-game, changing DS/WS screen rotation/layouts and similar, where squeezing everything into the same resolution is undesirable. It's also just more convenient than figuring out a resolution in the right aspect ratio and entering that in the display settings all the time.

Check if completed:
- [x] I have run any relevant test suites
- [x] I, the committer, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-06-22) and am compliant

resolves #1560, #3039, and #3410

### CODE REVIEW

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.Events.cs:618]: Did you mean
```suggestion
			foreach (var item in WindowSizeSubMenu.DropDownItems.Cast<ToolStripMenuItem>())
```
?
You shouldn't have to change it though.

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.Events.cs:634]: I think if you inserted this method above the other one the diff would be cleaner? Not that important

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.cs:63]: Unrelated removal of this local method?

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.cs:618]: Was it necessary to make these nullable to implement this new feature? If you could continue using `-1` as the sentinel value instead of `null`, I think that would be preferable. Otherwise, try `Point?` instead of 2 `int?`s.

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.cs:1378]: Omit docs and default argument on implementation

**Review by YoshiRulz — DISMISSED** [review]

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.Events.cs:618]: No, it's filtering out the separator

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:618]: It's not necessary to the PR, no. I just made this change because negative values including -1 are valid window positions

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (YoshiRulz) [reply src/BizHawk.Client.EmuHawk/MainForm.Events.cs:618]: I think the type check on `Tag` is enough for that, but you could also do something like:
```csharp
foreach (var item in WindowSizeSubMenu.DropDownItems.Cast<ToolStripItem>())
{
	if (item is not ToolStripSeparator && item.Tag is int itemScale)
	{
		// ...
	}
}
```
The confusing type hierarchy means it's not clear what `.OfType<ToolStripMenuItem>()` does just from looking at it; normally it'd be fine.

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:63]: I had to change this method to insert the items at the top of the menu (above the separator and new item added by this PR), and since there's no `InsertRange` method or similar the simplest way was inserting the items directly and dropping the whole array -> `AddRange` step.

Since theres nothing to return, the local method would just be

```csharp
void CreateWindowSizeFactorSubmenus() { ... }
CreateWindowSizeFactorSubmenus();
```

Comment (YoshiRulz) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:63]: Oh right. I suggest you add those new items programmatically rather than in the Designer. Alternatively, just add a "scoping block" around this bit and re-indent it, that will make the diff readable.

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.Events.cs:618]: The items need to be cast to `ToolStripMenuItem` to access the `Checked` property so there needs to be some kind of additional cast. I can change it to `item is ToolStripMenuItem menuItem` or something inside the loop if that's clearer

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (YoshiRulz) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:618]: Possibly related to https://github.com/TASEmulators/BizHawk/issues/3544?

Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:618]: Not directly, tool window position saving is handled elsewhere

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (kalimag) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:1378]: Are you sure? The majority of calls are from within `MainForm` itself, not through the interfaces

File: src/BizHawk.Client.EmuHawk/MainForm.Events.cs
Comment (YoshiRulz) [reply src/BizHawk.Client.EmuHawk/MainForm.Events.cs:618]: Yes, do that

File: src/BizHawk.Client.EmuHawk/MainForm.cs
Comment (YoshiRulz) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:1369]: IIRC if you omit the default argument you're sometimes forced to do `((IIntf) this).FrameBufferResized()` nonsense. I'm of the opinion that it should be explicit at every callsite anyway, but for this PR having it duplicated isn't a problem, it'll get fixed later.

Definitely drop the doc comment though.

**Comment by YoshiRulz** [comment]
Adding
```diff
 typeof(object),
+typeof(Point),
 typeof(Queue<>),
 typeof(/*Single*/float),
+typeof(Size),
 typeof(string),
```
to
https://github.com/TASEmulators/BizHawk/blob/dc582a4e7a1bc688a6eb6fdf8427fb9736bf10c9/src/BizHawk.Tests/Client.Common/config/SerializationStabilityTests.cs#L34-L37
should fix the test failure.

File: src/BizHawk.Client.Common/config/Config.cs
Comment (YoshiRulz) [reply src/BizHawk.Client.Common/config/Config.cs:141]: snipe'd :P

**Comment by kalimag** [comment]
`Point` and `Size` serialization will have to be re-tested on #3932

**Comment by kalimag** [comment]
> LGTM now, so I've made a note to merge this before the upcoming RC.
> 
> Thanks for all the contributions recently.

Well, I finally had to stop procrastinating to get any of this into a release before 2.10.1 in December 2025 ;)

**Comment by kalimag** [comment]
I have a few more changes that I didn't PR yet because of the merge conflicts with(out) this PR: https://github.com/kalimag/BizHawk/compare/pr/disable-automatic-resize...kalimag:BizHawk:pr/disable-automatic-resize-plus

Should I wait until this one is merged and open a new PR, or just push them into this PR?

**Comment by YoshiRulz** [comment]
I don't follow. Merge conflicts with `master`? You're welcome to rewrite history and force push, just comment with what you changed compared to the old HEAD.

**Comment by kalimag** [comment]
What I mean is that those changes are not strictly related to this PR, but the commits touch the same code and are based on this PR's branch rather than master.

So I can't trivially open a separate PR with just those changes without (probably) creating a bunch of merge conflicts between the two, and the easier option is to either have one PR (this one) containing all changes, or wait until this one is merged to open a second PR.

**Comment by skywind3000** [comment]
Thanks, but I still can't find the new option to keep the window size:

图片

I am using the latest `BizHawk-2.10-rc1-win-x64.zip`

**Comment by YoshiRulz** [comment]
`View` > `Window Size` > `Static Size`

**Comment by skywind3000** [comment]
thanks, that’s really helpful. Why not make it enable by default?

**Pull request merged**
